### ISSUE TITLE
Exception when using System.cmd inside a plug function

### ISSUE THREAD

**Issue by philipgiuliani** [mtrudel/bandit#18]
After adding `System.cmd("diff", ["file1.txt", "file2.txt"])` inside my handler I get the following exception:

```elixir
16:49:36.864 [error] GenServer #PID<0.509.0> terminating
** (FunctionClauseError) no function clause matching in Bandit.HTTP1.Handler.handle_info/2
    (bandit 0.4.5) lib/thousand_island/handler.ex:5: Bandit.HTTP1.Handler.handle_info({:EXIT, #Port<0.35>, :normal}, {%ThousandIsland.Socket{acceptor_id: "3162ABC96567", connection_id: "7278E72B3A7D", socket: #Port<0.34>, transport_module: ThousandIsland.Transports.TCP}, %{handler_module: Bandit.HTTP1.Handler, plug: {ShowroomWeb.Router, []}, read_timeout: 60000}})
    (stdlib 3.16.1) gen_server.erl:695: :gen_server.try_dispatch/4
    (stdlib 3.16.1) gen_server.erl:771: :gen_server.handle_msg/6
    (stdlib 3.16.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: {:EXIT, #Port<0.35>, :normal}
State: {%ThousandIsland.Socket{acceptor_id: "3162ABC96567", connection_id: "7278E72B3A7D", socket: #Port<0.34>, transport_module: ThousandIsland.Transports.TCP}, %{handler_module: Bandit.HTTP1.Handler, plug: {ShowroomWeb.Router, []}, read_timeout: 60000}}
```

I tried switching to Cowboy and there it works fine. Seems the process sends an exit message which is not handled properly.

**Comment by mtrudel** [comment]
Your hunch is correct, the issue here is that the process backing your HTTP connection is being notified that the process backing the `cmd` call has exited (Thousand Island handlers trap exit by design, which means that they'll see `:EXIT` messages as above instead of themselves terminating when a child process dies). 

Unfortunately, the Plug API doesn't provide a way to surface these into your plug implementation, so whatever path we take here would be policy on the part of Bandit, and might possibly result in opaque behaviour from the perspective of a plug implementor (for example, should we only trap 'normal' exits? Do we log / raise / silently drop errors otherwise?). It's unclear that there is an option here which would satisfy all users.

From a more philosophical perspective, I also think that such process details don't really belong on the user side of the Plug API, as they're really a backing detail of however Bandit decides to implement Plugs (for example, the issue you identify doesn't occur on HTTP2 connections since the HTTP2 implementation in Bandit explicitly handles process exits internally as part of how its process model works). The Plug API is simple, and part of that simplicity is that it is process agnostic; if you need to do things in your plug implementation which are aware of processes, I think the Plug answer to that is that you'll need to handle them yourself using a standalone supervision tree & process model. This also echoes generally accepted practice in Phoenix applications whereby your `*Web` application is best structured as a thin HTTP-mediated wrapper around your application's actual logic in a separate app.

Having said all of this, I think handling the exit message explicitly in the HTTP1 implementation makes sense, in order to bring it in line with HTTP2. On the one hand, part of Bandit's ethos is to express as little policy internally as possible, and any choice here is going to run counter to that. But on the other hand, the fact that HTTP2 already expresses a policy here (albeit unintentionally) makes us internally inconsistent. 

Fixed in 094d3c58f5bc3ce2ab90442be2ba7348d918e6b3.

**Comment by stevensonmt** [comment]
Do you really want to ignore the reason for the `:EXIT` message entirely? Is there no value to logging somehow to help with debugging in case the system process is not exiting normally?

**Comment by mtrudel** [comment]
@stevensonmt that's a very good question, and one that I deliberately didn't answer in this PR. As I mentioned above, the dogmatic side of me thinks that 'process messages are none of your business as a Plug user', which is correct from an API separation perspective (I *think*), but obviously deficient from a visibility & debugging perspective (as you point out, Bandit silently eating EXIT messages is the sort of thing that might cost someone hours of work to track down). 

As there is no way in the Plug API to hand such termination messages over to the Plug implementor, the best we can do in this case is to log. I would propose:

1. The clause added in 094d3c58f5bc3ce2ab90442be2ba7348d918e6b3 be refined to consider only normal termination (and to remain silent in this case).
2. A catchall `handle_info/2` handler be added to the end of the HTTP1 handler, which will catch abnormal termination in addition to any other unexpected messages (which will currently crash the handler). This will log at `warn`.
3. A similar catchall `handle_info/2` be added to the end of the HTTP2 handler. 
4. The existing `handle_info(:EXIT)` function in the HTTP2 handler be refined to only be silent in the case of Bandit 'internal' (aka stream process) pids terminating. Should a non-internal pid show up here, it will be logged.

Does this seem comprehensive @stevensonmt @philipgiuliani ?

**Comment by stevensonmt** [comment]
I think that sounds great.  While it might not be in keeping with the pure philosophical foundation of the library, I think users of the library will appreciate the added logging.

**Comment by mtrudel** [comment]
Midway through working this up, and I think I'm changing my tune a little bit.

I think the correct thing to do here is to have both the HTTP1 and HTTP2 implementations silently eat `:normal` exit messages, and explicitly NOT handle anything else. My reasoning is twofold:

* If we add a catchall `handle_info/2` call to log & notify of all messages sent, we should probably also do something similar for `handle_call/3` and `handle_cast/2` calls for consistency (that is, if we're making it so that unexpected messages do not crash a handler, we should do so for all message types). This seems to me to be solving the problem in the wrong place - those messages shouldn't be being sent in the first place. Presumably these messages are being sent for a reason, and since we don't have a way of passing them up to the user we should be complaining loudly and fatally that these messages are being sent to the 'wrong' place.

* If we were not trapping exits, the behaviour of abnormal termination would be to bring the handler process down; if we cared enough about this to handle it somehow, we would normally do so with a Supervisor mediated process tree. Point being: abnormal process termination isn't something we should live through, even if we log. We're essentially building an ad hoc, informally specified Supervisor implementation, even worse since all we can do is log it and not allow the user to take any action. I think it's much better in this case to not paper any of this over, and to blow up loudly. If the user needs to be able to handle these case, then they'll need to do it in their own process tree and leave Bandit to handle just HTTP concerns.

Based on this, my revised proposal is to:

* Have the HTTP1 handler silently accept `:normal` termination messages
* Not handle anything else (this means that anything else will die with a `GenServer termination` message)
* Have the HTTP2 handler continue to internally handle any termination messages from Bandit internal processes as it does today
* Have the HTTP2 handler silently accept `:normal` termination messages from external processes
* Not handle anything else (this means that anything else will die with a `GenServer termination` message)

What do you think?

**Comment by stevensonmt** [comment]
That's a reasonable approach. I guess the GenServer termination message will include the last message that killed it, providing a sort of de facto log for debugging anyway. I hadn't really been thinking about that earlier, but it's exactly what led OP to identify the problem. Not dying on normal exit is probably all that should be expected in that case.

**Comment by mtrudel** [comment]
Fixed in 0.4.7!
